Python/compile.c

```diff
@@ -9731,32 +9731,113 @@ optimize_cfg(cfg_builder *g, PyObject *consts, PyObject *const_cache)
         for (int i = 0; i < b->b_iused; i++) {
-            if ((b->b_instr[i].i_opcode == LOAD_CONST ||
-                b->b_instr[i].i_opcode == KW_NAMES) &&
-                    b->b_instr[i].i_oparg > max_const_index) {
-                max_const_index = b->b_instr[i].i_oparg;
+            if (b->b_instr[i].i_opcode == LOAD_CONST ||
+                b->b_instr[i].i_opcode == KW_NAMES) {
+
+                int index = b->b_instr[i].i_oparg;
+                index_map[index] = index;
             }
         }
     }
-    if (max_const_index+1 < PyList_GET_SIZE(consts)) {
-        //fprintf(stderr, "removing trailing consts: max=%d, size=%d\n",
-        //        max_const_index, (int)PyList_GET_SIZE(consts));
-        if (PyList_SetSlice(consts, max_const_index+1,
-                            PyList_GET_SIZE(consts), NULL) < 0) {
-            return 1;
+    /* now index_map[i] == i if consts[i] is used, -1 otherwise */
+
+    /* condense consts */
+    int n_used_consts = 0;
+    for (int i = 0; i < nconsts; i++) {
+        if (index_map[i] != -1) {
+            assert(index_map[i] == i);
+            if (n_used_consts != i) {
+                index_map[i] = -1;
+                index_map[n_used_consts] = i;
+            }
+            n_used_consts++;
         }
     }
-    return 0;
+    if (n_used_consts == nconsts) {
+        /* nothing to do */
+        err = 0;
+        goto end;
+    }
+
+    /* move all used consts to the beginning of the consts list */
+    for (Py_ssize_t i = 0; i < n_used_consts; i++) {
+        int old_index = index_map[i];
+        assert(i <= old_index);
+        if (i != old_index) {
+            PyObject *value = PyList_GET_ITEM(consts, index_map[i]);
+            assert(value != NULL);
+            int res = PyList_SetItem(consts, i, Py_NewRef(value));
+            assert(res == 0);
+        }
+    }
+
+    /* truncate the consts list at its new size */
+    if (PyList_SetSlice(consts, n_used_consts, nconsts, NULL) < 0) {
+        goto end;
+    }
+
+    /* adjust const indices in the bytecode */
+    reverse_index_map = PyMem_Malloc(nconsts * sizeof(int));
+    if (reverse_index_map == NULL) {
+        goto end;
+    }
+    for (int i = 0; i < nconsts; i++) {
+        reverse_index_map[i] = -1;
+    }
+    for (int i = 0; i < nconsts; i++) {
+        if (index_map[i] != -1) {
+            assert(reverse_index_map[index_map[i]] == -1);
+            reverse_index_map[index_map[i]] = i;
+        }
+    }
+
+    for (basicblock *b = entryblock; b != NULL; b = b->b_next) {
+        for (int i = 0; i < b->b_iused; i++) {
+            if (b->b_instr[i].i_opcode == LOAD_CONST ||
+                b->b_instr[i].i_opcode == KW_NAMES) {
+
+                int index = b->b_instr[i].i_oparg;
+                b->b_instr[i].i_oparg = reverse_index_map[index];
+            }
+        }
+    }
+
+    err = 0;
+end:
+    if (index_map != NULL) {
+        PyMem_Free(index_map);
```

Comment:
Doesn't `PyMem_Free()` do nothing if the value is `NULL`?

Reply:
Probably, I'll check when I sort out the other issues.